homeassistant/components/microbot_push/__init__.py

```diff
@@ -0,0 +1,146 @@
+    conf_dir = hass.config.path()
+    conf = (
+        conf_dir
+        + "/.storage/microbot-"
+        + re.sub("[^a-f0-9]", "", bdaddr.lower())
+        + ".conf"
+    )
+    client = MicroBotApiClient(
+        device=ble_device,
+        config=conf,
+        retry_count=DEFAULT_RETRY_COUNT,
+    )
+    coordinator = MicroBotDataUpdateCoordinator(
+        hass, client=client, ble_device=ble_device
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(
+        entry, PLATFORMS
+    )
+
+    async def generate_token(call: ServiceCall) -> None:
+        _LOGGER.debug("Token service called")
+        await coordinator.api.connect(init=True)
+
+    async def calibrate(call: ServiceCall) -> None:
+        _LOGGER.debug("Calibrate service called")
+        depth = call.data["depth"]
+        duration = call.data["duration"]
+        mode = call.data["mode"]
+        await coordinator.api.connect()
+        coordinator.api.setDepth(depth)
+        coordinator.api.setDuration(duration)
+        coordinator.api.setMode(mode)
+        await coordinator.api.calibrate()
+        await coordinator.api.disconnect()
+
+    hass.services.async_register(DOMAIN, "generate_token", generate_token)
+    hass.services.async_register(DOMAIN, "calibrate", calibrate)
+    entry.async_on_unload(entry.add_update_listener(async_reload_entry))
+    return True
+
+
+class MicroBotDataUpdateCoordinator(PassiveBluetoothDataUpdateCoordinator):
+    """Class to manage fetching data from the MicroBot."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: MicroBotApiClient,
+        ble_device: BLEDevice,
+    ) -> None:
+        """Initialize."""
+        self.api = client
+        self.platforms: list[str] = []
+        self._ready_event = asyncio.Event()
+        self.data: dict[str, Any] = {}
+        self.ble_device = ble_device
+
+        super().__init__(
+            hass, _LOGGER, ble_device.address, bluetooth.BluetoothScanningMode.ACTIVE
+        )
+
+    @callback
+    def _async_handle_bluetooth_event(
+        self,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Handle a Bluetooth event."""
+        super()._async_handle_bluetooth_event(service_info, change)
+        if adv := parse_advertisement_data(
+            service_info.device, service_info.advertisement
+        ):
+            self.data = adv.data
+            if self.data:
+                self._ready_event.set()
+            _LOGGER.debug("%s: MicroBot data: %s", self.ble_device.address, self.data)
+            self.api.update_from_advertisement(adv)
+        self.async_update_listeners()
+
+    async def async_wait_ready(self) -> bool:
+        """Wait for the device to be ready."""
+        try:
+            await asyncio.wait_for(self._ready_event.wait(), timeout=55)
```

Comment:
We should use `async_timeout.timeout` instead.

Reply:
It looks like this was copied from switchbot.

I'll fix that one https://github.com/home-assistant/core/pull/76630